src/storage/src/types/errors.rs

```diff
@@ -166,6 +344,7 @@ pub enum DataflowError {
     DecodeError(DecodeError),
     EvalError(EvalError),
     SourceError(SourceError),
+    EnvelopeError(EnvelopeError),
```

Comment:
Is this the structure we want? You may be alluding to this when you talk about the api breakage in your pr summary, but what if we used this structure:

```
pub enum DataflowError {
    Definite(DefiniteError)
    Indefinite(IndefiniteError)
}

pub enum DefiniteError {
    KeyDecode(DecodeErrorPlusVecU8Value),
    Value(UpsertValueError),
}

pub enum DefiniteError {
    DecodeError(DecodeError),
    EvalError(EvalError),
    SourceError(SourceError),
    EnvelopeError(EnvelopeError),
}

pub enum EnvelopeError {
    Debezium(String),
    Flat(String),
}
```

Moving things around depending on if they are definite. That way, the decode stage can produce both indefinite and definite errors, and upsert can pass through the indefinite ones, and produce any `DataflowErrow`

I could be missing something!

Reply:
This might work! But I wanted to add something minimally invasive now, because I believe in the future folks on the storage team will be thinking about how to best split up errors type, so doing the simplest possible thing now seemed like the best strategy for avoiding wasted work.

Reply:
Side note: I don't think `DataflowError` should have a variant for inedefinite errors. Those should not end up in the "Err" collections that we use to present errors to a user when a view is wedged. And indefinite errors should definitely (ha!) not make it into the storage collection. The main reason is that indefinite errors are not a function of the data, hence cannot be reproduced, hence can never be retracted.

Reply:
@aljoscha that is a good point, we probably need to do more invasive surgery to use the type system to prevent problems like your describe